File: Rust_Solutions/solution3.rs
use std::env;

// User input
fn find_max(a: i32, b: i32) -> i32 {
    if a > b { a } else { b }
}
// End of user input

struct Checker;

impl Checker {
    fn check(func: fn(i32, i32) -> i32, a: i32, b: i32, expected: i32) {
        let result = func(a, b);
        if result == expected {
            println!("Test Case Passed: Result = {}, Expected = {}", result, expected);
        } else {
            println!("Failed Test Case: Result = {}, Expected = {}", result, expected);
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!("Usage: {} <a> <b> <expected>", args[0]);
        std::process::exit(1);
    }

    let a: i32 = args[1].parse().expect("Invalid first number");
    let b: i32 = args[2].parse().expect("Invalid second number");
    let expected: i32 = args[3].parse().expect("Invalid expected value");

    Checker::check(find_max, a, b, expected);
} 

File: Solutions/Cpp_Solutions/solution3.cpp
#include <iostream>
#include <string>

// User input
int findMax(int a, int b) {
    return (a > b) ? a : b;
}
// End of user input

class Checker {
public:
    static void check(int (*func)(int, int), int a, int b, int expected) {
        int result = func(a, b);
        if (result == expected) {
            std::cout << "Test Case Passed: Result = " << result 
                      << ", Expected = " << expected << std::endl;
        } else {
            std::cout << "Failed Test Case: Result = " << result 
                      << ", Expected = " << expected << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 4) {
        std::cerr << "Usage: " << argv[0] << " <a> <b> <expected>" << std::endl;
        return 1;
    }

    int a = std::stoi(argv[1]);
    int b = std::stoi(argv[2]);
    int expected = std::stoi(argv[3]);

    Checker::check(findMax, a, b, expected);
    return 0;
} 